frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1456 @@
+		/// Maximum number of iteration of balancing that will be executed in the embedded miner of
+		/// the pallet.
+		type MinerMaxIterations: Get<u32>;
+		/// Maximum weight that the miner should consume.
+		///
+		/// The miner will ensure that the total weight of the unsigned solution will not exceed
+		/// this values, based on [`WeightInfo::submit_unsigned`].
+		type MinerMaxWeight: Get<Weight>;
+
+		/// Something that will provide the election data.
+		type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+
+		/// The compact solution type
+		type CompactSolution: codec::Codec
+			+ Default
+			+ PartialEq
+			+ Eq
+			+ Clone
+			+ sp_std::fmt::Debug
+			+ CompactSolution;
+
+		/// Accuracy used for fallback on-chain election.
+		type OnChainAccuracy: PerThing128;
+
+		/// Configuration for the fallback
+		type Fallback: Get<FallbackStrategy>;
+
+		/// The configuration of benchmarking.
+		type BenchmarkingConfig: BenchmarkingConfig;
+
+		/// The weight of the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let next_election = T::DataProvider::next_election_prediction(now).max(now);
+
+			let signed_deadline = T::SignedPhase::get() + T::UnsignedPhase::get();
+			let unsigned_deadline = T::UnsignedPhase::get();
+
+			let remaining = next_election - now;
+			let current_phase = Self::current_phase();
+
+			match current_phase {
+				Phase::Off if remaining <= signed_deadline && remaining > unsigned_deadline => {
+					Self::on_initialize_open_signed();
+					log!(info, "Starting signed phase at #{:?} , round {}.", now, Self::round());
+					T::WeightInfo::on_initialize_open_signed()
+				}
+				Phase::Signed | Phase::Off
+					if remaining <= unsigned_deadline && remaining > 0u32.into() =>
+				{
+					let (need_snapshot, enabled, additional) = if current_phase == Phase::Signed {
+						// followed by a signed phase: close the signed phase, no need for snapshot.
+						// TWO_PHASE_NOTE: later on once we have signed phase, this should return
+						// something else.
+						(false, true, Weight::zero())
+					} else {
+						// no signed phase: create a new snapshot, definitely `enable` the unsigned
+						// phase.
+						(true, true, Weight::zero())
+					};
+
+					Self::on_initialize_open_unsigned(need_snapshot, enabled, now);
+					log!(info, "Starting unsigned phase({}) at #{:?}.", enabled, now);
+
+					let base_weight = if need_snapshot {
+						T::WeightInfo::on_initialize_open_unsigned_with_snapshot()
+					} else {
+						T::WeightInfo::on_initialize_open_unsigned_without_snapshot()
+					};
+					base_weight.saturating_add(additional)
+				}
+				_ => T::WeightInfo::on_initialize_nothing(),
+			}
+		}
+
+		fn offchain_worker(n: T::BlockNumber) {
+			// We only run the OCW in the first block of the unsigned phase.
+			if Self::current_phase().is_unsigned_open_at(n) {
```

Comment:
I guess we could simply use `is_unsigned_open` here no ? because block number is the current block number not something in the future.
Also it seems Unsigned could refactored to get rid of `bool` and `block number`
we never have an unsigned which is false, and block number is never needed AFAICT.

Reply:
I think Phase should be refactored to:
```
/// Current phase of the pallet.
#[derive(PartialEq, Eq, Clone, Copy, Encode, Decode, RuntimeDebug)]
pub enum Phase<Bn> {
	Off,
	Signed,
	Unsigned,
}
```
I don't understand the need for the unsigned variant fields

Reply:
Both are for the sake of future proofing: 

- The `bool` is needed because we are likely to have conditional situations where we might open the unsigned phase based on how the signed phase went. 
- The `block_number` we can get rid of it, I honestly added it mostly for debugging purposes.

Reply:
OK, I see both fields are for future usage.